### DISCUSSION TITLE
Simulation computationally too expensive

### DISCUSSION THREAD

**Discussion by moritzboeker** [FieldRobotEvent/Virtual_Field_Robot_Event#24] [Q&A] [answered]
Dear FRE2021-Team,

thank you for the beautiful Gazebo-Simulation you've set up.
Unfortunately the simulation is now so detailed that we cannot use it anymore,
as its demand on computational power is too high.
I have assigned

- 16 GB RAM
- 3 GB of the Nvidia GPU Video RAM
-  8 cores

to the Ubuntu 18.04 LTS Virtual Machine (VMWare) and only yield about 0.5 to 1 FPS at the task 1 mini wolrd - and this is when our robot program isnt't running at all.
Here is an image if it helps:
https://ibb.co/YZBgKSF

Could you please think about making a more simpler version of the simulation? Or is there something from my side I could do to improve the simulation FPS?

Kind regards,

Moritz

**Comment by ThijsRuigrok** [comment, marked as answer]
Dear Mortiz,
Thank you for your question, and the screenshot. First thing I notice in your image is that the heightmap does not match the location of the plants. I suspect that Gazebo accidentally loaded an older heightmap with a larger resolution(also causing a slow simulation). This problem should be solved in the latest version of this repo.
There are several possibilities to speed up the simulation (none of them is perfect). For the simulation speed, the FPS count doesn’t matter. The FPS represent how fast the scene is rendered for the gui, and does not reflect the speed of the simulation. What matters is the Real-Time Factor. This is the rate at which the simulation is run with respect to the actual time. With a computer like yours, I expect(wild guess) a Real-Time Factor of around 0.5.
Possible speedups:
-	You are using VM. What hardware did you allocate to your VM? Sometimes, only one core and one GB of ram is allocated to the VM, limiting the computation power. You can change this in the VM-ware settings menu when the VM is powered off. I recommend allocating at least 8GB of RAM and all your 8 cores to the VM(your host system will share the cores, but not the RAM). 
-	GPU virtualization from an Nvidia GPU is not possible(this is blocked by NVidia). Therefore your VM cannot use the GPU to accelerate rendering. Ideally, you can install Ubuntu natively(as a dual boot) without using a VM. This will limit computational overhead and allow Gazebo to use your GPU.
-	You can launch Gazebo without gui, by using the command: roslaunch virtual_maize_field jackal_simulation.launch gui:=false. This will save some resources. 
-	In gazebo you can change the settings of the ODE physics calculator. Increasing the “max-step_size” will drastically increase the simulation speed. However, when this parameter becomes too large, the physics can become unstable. 
Please let me know if any of those solutions worked for you.

**Reply by moritzboeker** [reply]
Thank you so much, this helped a lot. Especially increasing the max_step_step brought me to a real time factor of more than 1.0. I think about setting up a dual boot since there is no nvidia GPU passthrough from a windows host to ubuntu virtual machine.

**Reply by Darkproduct** [reply]
I would suggest to setup dual-boot for the real simulation tests. Almost all my team members are using Linux anyway and we can run the simulation on worse hardware then yours. 

I tested VMware on my main system with an Ryzen 5600X, 32 GB DDR4 and GTX 680 with 2GB VRAM and get up to a real time factor of 0.95 with the default settings (but that's without the correct CPU laser output).

**Reply by moritzboeker** [reply]
Thank you Darkproduct, I now have installed Ubuntu as separate OS and now its working really smoothly!
